util/rowcodec/common.go

```diff
@@ -240,3 +245,109 @@ func IsNewFormat(rowData []byte) bool {
 func FieldTypeFromModelColumn(col *model.ColumnInfo) *types.FieldType {
 	return col.FieldType.Clone()
 }
+
+// ColData combines the column info as well as its datum. It's used to calculate checksum.
+type ColData struct {
+	*model.ColumnInfo
+	Datum *data.Datum
+}
+
+// Encode encodes the column datum into bytes for checksum. If buf provided, append encoded data to it.
+func (c ColData) Encode(buf []byte) ([]byte, error) {
+	return appendDatumForChecksum(buf, c.Datum, c.GetType())
+}
+
+// RowData is a list of ColData for row checksum calculation.
+type RowData struct {
+	Cols []ColData
+	Data []byte
+}
+
+// Len implements sort.Interface for RowData.
+func (r RowData) Len() int { return len(r.Cols) }
+
+// Less implements sort.Interface for RowData.
+func (r RowData) Less(i int, j int) bool { return r.Cols[i].ID < r.Cols[j].ID }
+
+// Swap implements sort.Interface for RowData.
+func (r RowData) Swap(i int, j int) { r.Cols[i], r.Cols[j] = r.Cols[j], r.Cols[i] }
```

Comment:
It looks the `Data` is not sorted

Reply:
`Data` here is just a buffer for encoding column, so no need to be sorted. It's named `Data` because `RowData.Encode` writes encoded data to it (so user can also access the result from `rd.Data`).